tests/lib/mirrors/reflected_type_generics_test.dart

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) 2013, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library test.reflected_type_generics_test;
+
+import 'dart:mirrors';
+
+import 'package:expect/expect.dart';
+
+import 'reflected_type_helper.dart';
+
+class A<T> {}
+class P {}
+class B extends A<P> {}
+class C<K, V> {}
+class D<T> extends A<T> {}
+class E<K> extends C<K, int> {}
+
+main() {
+  // "Happy" paths:
+  expectReflectedType(reflectType(A, [P]), new A<P>().runtimeType);
+  expectReflectedType(reflectType(C, [B, P]), new C<B, P>().runtimeType);
+  expectReflectedType(reflectType(D, [P]), new D<P>().runtimeType);
+  expectReflectedType(reflectType(E, [P]), new E<P>().runtimeType);
+
+  // Edge cases:
+  Expect.throws(
+    () => reflectType(P, []),
+    (e) => e is ArgumentError && e.invalidValue is List,
+    "Should throw an ArgumentError if reflecting not a generic class with empty list of type arguments");
```

Comment:
@floitschG @rmacnak-google  I've been trying to wrap my head around error handling in such cases.

Could you help me in understanding "best practices" here?

Basically, I'm currently using pretty verbose way to handle argument errors, e.g.:

```cpp
  if (!clz.IsGeneric()) {
    const Array& error_args = Array::Handle(Array::New(3));
    error_args.SetAt(0, type);
    error_args.SetAt(1, String::Handle(String::New("key")));
    error_args.SetAt(2, String::Handle(
      String::New("Type must be a generic class.")));
    Exceptions::ThrowByType(Exceptions::kArgumentValue, error_args);
    UNREACHABLE();
  }
```

That's why I put these strict assertions here so that I can actually verify my code.
However as @floitschG pointed out this is undesirable and we only need to know that an `Error` happened.

Now, I realize that I can replace those verbose `if` statements with a call to `ASSERT`, so that it would become:

```cpp
ASSERT(clz.IsGeneric());
// or:
ASSERT(clz.NumTypeParameters() != args.Length());
```

This would throw a `RuntimeError`. I see that this approach is used a lot throughout the codebase, while constructing exceptions like I do currently is fairly rare.

Would you say that `ASSERT` is preferable in such cases?

I feel like providing `ArgumentError` with detailed context is more user-friendly, but on the other hand it's really verbose and maybe provided value is not justified in terms of maintenance cost.